src/pal/src/sharedmemory/sharedmemory.cpp

```diff
@@ -0,0 +1,1114 @@
+#endif // _DEBUG
+}
+
+void SharedMemoryManager::ReleaseCreationDeletionProcessLock()
+{
+    _ASSERTE(IsCreationDeletionProcessLockAcquired());
+    _ASSERTE(!IsCreationDeletionFileLockAcquired());
+
+#ifdef _DEBUG
+    s_creationDeletionProcessLockOwnerThreadId = SharedMemoryHelpers::InvalidPid;
+#endif // _DEBUG
+    LeaveCriticalSection(&s_creationDeletionProcessLock);
+}
+
+void SharedMemoryManager::AcquireCreationDeletionFileLock()
+{
+    _ASSERTE(IsCreationDeletionProcessLockAcquired());
+    _ASSERTE(!IsCreationDeletionFileLockAcquired());
+
+    if (s_creationDeletionLockFileDescriptor == -1)
+    {
+        if (!SharedMemoryHelpers::EnsureDirectoryExists(SHARED_MEMORY_TEMP_DIRECTORY_PATH, false /* createIfNotExist */))
+        {
+            throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::IO));
+        }
+        SharedMemoryHelpers::EnsureDirectoryExists(SHARED_MEMORY_RUNTIME_TEMP_DIRECTORY_PATH);
+        SharedMemoryHelpers::EnsureDirectoryExists(SHARED_MEMORY_SHARED_MEMORY_DIRECTORY_PATH);
+        s_creationDeletionLockFileDescriptor = SharedMemoryHelpers::OpenDirectory(SHARED_MEMORY_SHARED_MEMORY_DIRECTORY_PATH);
+        if (s_creationDeletionLockFileDescriptor == -1)
+        {
+            throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::IO));
+        }
+    }
+
+    bool acquiredFileLock = SharedMemoryHelpers::TryAcquireFileLock(s_creationDeletionLockFileDescriptor, LOCK_EX);
+    _ASSERTE(acquiredFileLock);
+#ifdef _DEBUG
+    s_creationDeletionFileLockOwnerThreadId = GetCurrentThreadId();
+#endif // _DEBUG
+}
+
+void SharedMemoryManager::ReleaseCreationDeletionFileLock()
+{
+    _ASSERTE(IsCreationDeletionProcessLockAcquired());
+    _ASSERTE(IsCreationDeletionFileLockAcquired());
+    _ASSERTE(s_creationDeletionLockFileDescriptor != -1);
+
+#ifdef _DEBUG
+    s_creationDeletionFileLockOwnerThreadId = SharedMemoryHelpers::InvalidPid;
+#endif // _DEBUG
+    SharedMemoryHelpers::ReleaseFileLock(s_creationDeletionLockFileDescriptor);
+}
+
+#ifdef _DEBUG
+bool SharedMemoryManager::IsCreationDeletionProcessLockAcquired()
+{
+    return s_creationDeletionProcessLockOwnerThreadId == GetCurrentThreadId();
+}
+
+bool SharedMemoryManager::IsCreationDeletionFileLockAcquired()
+{
+    return s_creationDeletionFileLockOwnerThreadId == GetCurrentThreadId();
+}
+#endif // _DEBUG
+
+void SharedMemoryManager::AddProcessDataHeader(SharedMemoryProcessDataHeader *processDataHeader)
+{
+    _ASSERTE(processDataHeader != nullptr);
+    _ASSERTE(IsCreationDeletionProcessLockAcquired());
+    _ASSERTE(processDataHeader->GetNextInProcessDataHeaderList() == nullptr);
+    _ASSERTE(FindProcessDataHeader(processDataHeader->GetId()) == nullptr);
+
+    processDataHeader->SetNextInProcessDataHeaderList(s_processDataHeaderListHead);
+    s_processDataHeaderListHead = processDataHeader;
+}
+
+void SharedMemoryManager::RemoveProcessDataHeader(SharedMemoryProcessDataHeader *processDataHeader)
```

Comment:
I actually need a hash table here, since I also need to find an item in O(1) time by name. I have a TODO in the Find function regarding this. I need to go through all the various hash tables we have and see if I can move one to be usable in PAL, or try to use the STL one. Figured I'd do that later since it's not a big deal for now.